### PR TITLE
Fallback sum/add to CPU if needed

### PR DESCRIPTION
**Pull request by gujinghui**
Fallback sum/add to CPU if needed

### CODE REVIEW

**Comment by yinghai** [comment]
Can we put in comments in the code to clearly state under what condition we need to fall back to CPU? And what's the use case for this?

**Comment by gujinghui** [comment]
@yinghai
rebased, added unit test.
depends on #15726

**Comment by gujinghui** [comment]
case failed because of issues fixed in #15726 

> Jan 04 07:58:05 ../../../../usr/local/caffe2/lib/python2.7/dist-packages/caffe2/python/ideep/elementwise_sum_op_test.py::ElementwiseSumTest::test_elementwise_sum_fallback ./.jenkins/caffe2/test.sh: line 96:  1189 Segmentation fault      (core dumped) "$PYTHON" -m pytest -x -v --disable-warnings --junit-xml="$pytest_reports_dir/result.xml" --ignore "$CAFFE2_PYPATH/python/test/executor_test.py" --ignore "$CAFFE2_PYPATH/python/operator_test/matmul_op_test.py" --ignore "$CAFFE2_PYPATH/python/operator_test/pack_ops_test.py" --ignore "$CAFFE2_PYPATH/python/mkl/mkl_sbn_speed_test.py" ${rocm_ignore_test[@]} "$CAFFE2_PYPATH/python" "${EXTRA_TESTS[@]}"
Exited with code 139

**Review by yinghai — CHANGES_REQUESTED** [review]
The following test is failing:  caffe2/python/ideep/elementwise_sum_op_test.py::ElementwiseSumTest::test_elementwise_sum_fallback

**Comment by gujinghui** [comment]
@yinghai 
the case failures are caused by issues fixed in #15726.
Pls merge #15726, then I will rebase this PR.

**Comment by gujinghui** [comment]
@yinghai 
rebased. pls help review.

**Comment by gujinghui** [comment]
@yinghai 

The failures should be not caused by this PR.

1. 
> update-initramfs: Generating /boot/initrd.img-4.4.0-96-generic
/usr/local/lib/python2.7/dist-packages/urllib3/util/ssl_.py:369: SNIMissingWarning: An HTTPS request has been made, but the SNI (Server Name Indication) extension to TLS is not available on this platform. This may cause the server to present an incorrect TLS certificate, which can cause validation failures. You can upgrade to a newer version of Python to solve this. For more information, see https://urllib3.readthedocs.io/en/latest/advanced-usage.html#ssl-warnings
  SNIMissingWarning
WARNING! Using --password via the CLI is insecure. Use --password-stdin.
Error response from daemon: Get https://308535385114.dkr.ecr.us-east-1.amazonaws.com/v2/: x509: certificate signed by unknown authority (possibly because of "crypto/rsa: verification error" while trying to verify candidate authority certificate "Amazon")
Exited with code 1

2.
> Jan 11 07:39:25 Downloading: "https://download.pytorch.org/models/resnet18-5c106cde.pth" to /var/lib/jenkins/.torch/models/resnet18-5c106cde.pth
Jan 11 07:39:25 ERROR
Jan 11 07:39:25 test_check_onnx_broadcast (__main__.TestONNXUtils) ... ok
Jan 11 07:39:25 test_prepare_onnx_paddings (__main__.TestONNXUtils) ... ok
Jan 11 07:39:25 
Jan 11 07:39:25 ======================================================================
Jan 11 07:39:25 ERROR: setUpClass (__main__.TestHub)
Jan 11 07:39:25 ----------------------------------------------------------------------
Jan 11 07:39:25 Traceback (most recent call last):
Jan 11 07:39:25   File "/opt/conda/lib/python3.6/site-packages/urllib3/connection.py", line 159, in _new_conn
Jan 11 07:39:25     (self._dns_host, self.port), self.timeout, **extra_kw)
Jan 11 07:39:25   File "/opt/conda/lib/python3.6/site-packages/urllib3/util/connection.py", line 57, in create_connection
Jan 11 07:39:25     for res in socket.getaddrinfo(host, port, family, socket.SOCK_STREAM):
Jan 11 07:39:25   File "/opt/conda/lib/python3.6/socket.py", line 745, in getaddrinfo
Jan 11 07:39:25     for res in _socket.getaddrinfo(host, port, family, type, proto, flags):
Jan 11 07:39:25 socket.gaierror: [Errno -3] Temporary failure in name resolution

3.
> 06:48:30 ../../../../usr/local/caffe2/lib/python2.7/dist-packages/caffe2/python/operator_test/fc_operator_test.py::TestFcOperator::test_fc_transposed ++ echo 'Stopping container...'
06:48:30 Stopping container...
06:48:30 ++ '[' -n '' ']'
06:48:30 ++ docker rm -f d228611ed513bfcee022aa9550064b0570b04299db705a48e6e2c04ad1a4f939
06:48:30 FATAL: command execution failed
06:48:30 java.nio.channels.ClosedChannelException
06:48:30 	at org.jenkinsci.remoting.protocol.NetworkLayer.onRecvClosed(NetworkLayer.java:154)
06:48:30 	at org.jenkinsci.remoting.protocol.impl.NIONetworkLayer.ready(NIONetworkLayer.java:142)
06:48:30 	at org.jenkinsci.remoting.protocol.IOHub$OnReady.run(IOHub.java:789)
06:48:30 	at jenkins.util.ContextResettingExecutorService$1.run(ContextResettingExecutorService.java:28)
06:48:30 	at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1149)
06:48:30 	at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:624)
06:48:30 	at java.lang.Thread.run(Thread.java:748)
06:48:30 Caused: java.io.IOException: Backing channel 'JNLP4-connect connection from 205.234.28.201/205.234.28.201:26927' is disconnected.
06:48:30 	at hudson.remoting.RemoteInvocationHandler.channelOrFail(RemoteInvocationHandler.java:209)
06:48:30 	at hudson.remoting.RemoteInvocationHandler.invoke(RemoteInvocationHandler.java:278)
06:48:30 	at com.sun.proxy.$Proxy84.isAlive(Unknown Source)
06:48:30 	at hudson.Launcher$RemoteLauncher$ProcImpl.isAlive(Launcher.java:1137)
06:48:30 	at hudson.Launcher$RemoteLauncher$ProcImpl.join(Launcher.java:1129)
06:48:30 	at hudson.tasks.CommandInterpreter.join(CommandInterpreter.java:155)
06:48:30 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:109)
06:48:30 	at hudson.tasks.CommandInterpreter.perform(CommandInterpreter.java:66)
06:48:30 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
06:48:30 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
06:48:30 	at hudson.model.Build$BuildExecution.build(Build.java:206)
06:48:30 	at hudson.model.Build$BuildExecution.doRun(Build.java:163)
06:48:30 	at hudson.model.AbstractBuild$AbstractBuildExecution.run(AbstractBuild.java:504)
06:48:30 	at hudson.model.Run.execute(Run.java:1727)
06:48:30 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
06:48:30 	at hudson.model.ResourceController.execute(ResourceController.java:97)
06:48:30 	at hudson.model.Executor.run(Executor.java:429)
06:48:30 Build step 'Execute shell' marked build as failure
06:48:30 [xUnit] [INFO] - Starting to record.
06:48:30 [xUnit] [INFO] - Processing GoogleTest-1.6
06:48:30 ERROR: Build step failed with exception
06:48:30 java.lang.NullPointerException
06:48:30 	at org.jenkinsci.plugins.xunit.XUnitProcessor.performTests(XUnitProcessor.java:145)
06:48:30 	at org.jenkinsci.plugins.xunit.XUnitProcessor.performXUnit(XUnitProcessor.java:88)
06:48:30 	at org.jenkinsci.plugins.xunit.XUnitPublisher.perform(XUnitPublisher.java:142)
06:48:30 	at org.jenkinsci.plugins.xunit.XUnitPublisher.perform(XUnitPublisher.java:134)
06:48:30 	at hudson.tasks.BuildStepMonitor$1.perform(BuildStepMonitor.java:20)
06:48:30 	at hudson.model.AbstractBuild$AbstractBuildExecution.perform(AbstractBuild.java:744)
06:48:30 	at hudson.model.AbstractBuild$AbstractBuildExecution.performAllBuildSteps(AbstractBuild.java:690)
06:48:30 	at hudson.model.Build$BuildExecution.post2(Build.java:186)
06:48:30 	at hudson.model.AbstractBuild$AbstractBuildExecution.post(AbstractBuild.java:635)
06:48:30 	at hudson.model.Run.execute(Run.java:1752)
06:48:30 	at hudson.model.FreeStyleBuild.run(FreeStyleBuild.java:43)
06:48:30 	at hudson.model.ResourceController.execute(ResourceController.java:97)
06:48:30 	at hudson.model.Executor.run(Executor.java:429)
06:48:30 Build step 'Publish xUnit test result report' marked build as failure
06:48:30 Finished: FAILURE

**Comment by gujinghui** [comment]
the failures should be not related to this PR

> Jan 24 04:57:56 test_single_tensor (__main__.TestTensorDataset) ... ok
Jan 24 04:57:56 
Jan 24 04:57:56 ======================================================================
Jan 24 04:57:56 FAIL: test_proper_exit (__main__.TestDataLoader)
Jan 24 04:57:56 There might be ConnectionResetError or leaked semaphore warning (due to dirty process exit), but they are all safe to ignore
Jan 24 04:57:56 ----------------------------------------------------------------------
Jan 24 04:57:56 Traceback (most recent call last):
Jan 24 04:57:56   File "/var/lib/jenkins/workspace/test/common_utils.py", line 120, in wrapper
Jan 24 04:57:56     fn(*args, **kwargs)
Jan 24 04:57:56   File "test_dataloader.py", line 785, in test_proper_exit
Jan 24 04:57:56     self.assertIsInstance(loader_p.exception, RuntimeError, desc)
Jan 24 04:57:56 AssertionError: ConnectionRefusedError('Traceback (most recent call last):\n  File "/var/lib/jenkins/workspace/test/test_dataloader.py", line 175, in run\n    super(ErrorTrackingProcess, self).run()\n  File "/opt/conda/lib/python3.6/multiprocessing/process.py", line 93, in run\n    self._target(*self._args, **self._kwargs)\n  File "/var/lib/jenkins/workspace/test/test_dataloader.py", line 354, in _test_proper_exit\n    for i, _ in enumerate(it):\n  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 514, in __next__\n    idx, batch = self._get_batch()\n  File "/opt/conda/lib/python3.6/site-packages/torch/utils/data/dataloader.py", line 493, in _get_batch\n    return self.data_queue.get()\n  File "/opt/conda/lib/python3.6/multiprocessing/queues.py", line 113, in get\n    return _ForkingPickler.loads(res)\n  File "/opt/conda/lib/python3.6/site-packages/torch/multiprocessing/reductions.py", line 257, in rebuild_storage_fd\n    fd = df.detach()\n  File "/opt/conda/lib/python3.6/multiprocessing/resource_sharer.py", line 57, in detach\n    with _resource_sharer.get_connection(self._id) as conn:\n  File "/opt/conda/lib/python3.6/multiprocessing/resource_sharer.py", line 87, in get_connection\n    c = Client(address, authkey=process.current_process().authkey)\n  File "/opt/conda/lib/python3.6/multiprocessing/connection.py", line 487, in Client\n    c = SocketClient(address)\n  File "/opt/conda/lib/python3.6/multiprocessing/connection.py", line 614, in SocketClient\n    s.connect(address)\nConnectionRefusedError: [Errno 111] Connection refused\n',) is not an instance of <class 'RuntimeError'> : test_proper_exit with use_workers=True, pin_memory=False, hold_iter_reference=True, exit_method=worker_kill
Jan 24 04:57:56 
Jan 24 04:57:56 ----------------------------------------------------------------------
Jan 24 04:57:56 Ran 48 tests in 11.791s
Jan 24 04:57:56 
Jan 24 04:57:56 FAILED (failures=1, skipped=5)
Jan 24 04:57:56 Traceback (most recent call last):
Jan 24 04:57:56   File "test/run_test.py", line 455, in <module>
Jan 24 04:57:56     main()
Jan 24 04:57:56   File "test/run_test.py", line 447, in main
Jan 24 04:57:56     raise RuntimeError(message)
Jan 24 04:57:56 RuntimeError: test_dataloader failed!
Jan 24 04:57:56 + cleanup
Jan 24 04:57:56 + retcode=1
Jan 24 04:57:56 + set +x

**Comment by gujinghui** [comment]
@yinghai
rebased. pls help review.

**Comment by gujinghui** [comment]
@yinghai
rebased. pls help review.

**Comment by gujinghui** [comment]
@yinghai
rebased. pls help review.

File: caffe2/ideep/operators/elementwise_sum_op.cc [hunk 1]
```diff
@@ -1,44 +1,73 @@
...
+        ideep::direct_copy::compute(X, *Y);
+      } else {
+        const vector<float> scales(InputSize(), 1.0);
+        ideep::sum::compute(scales, inputs_itensor, *Y);
+      }
+      return true;
     }
 
-    return true;
+    if (InputSize() == 2)
```
Comment (yinghai) [inline caffe2/ideep/operators/elementwise_sum_op.cc]: Add {}

**Review by yinghai — CHANGES_REQUESTED** [review]
LGTM. Please address comments.

**Comment by gujinghui** [comment]
rebased. pls help review. @yinghai

**Review by yinghai — APPROVED** [review]

**Pull request closed without merging**